### PR TITLE
secrets/database: advanced TTL management for static roles

### PR DESCRIPTION
**Pull request by fairclothjm**
This PR enables schedule-based auto-rotation of static roles for the Database Secrets engine. This rotation mechanism will be “embedded” in the Database Secrets engine and will not be reusable by other engines. In particular, this approach allows for exclusion windows (e.g. market hours) in which automatic rotations will not occur.

We add a new field `rotation_schedule` to the [create static role](https://developer.hashicorp.com/vault/api-docs/secret/databases#create-static-role) endpoint. This field can be set with a cron-style string that will define the schedule on which rotations should occur:

```
// set rotations to occur on Saturday at 00:00
vault write database/static-roles/test-role \
    db_name=postgresql \
    username="vault" \
    rotation_schedule="0 0 * * SAT"
```

Additionally, we add a new `rotation_window` field on the [create static role](https://developer.hashicorp.com/vault/api-docs/secret/databases#create-static-role) endpoint. It will be optional when `rotation_schedule` is set and disallowed when `rotation_period` is set. This field can be set with a [time duration string](https://developer.hashicorp.com/vault/docs/concepts/duration-format) with a minimum of 1 hour. The `rotation_window` will define the window of time in which rotations are allowed to occur starting from a given `rotation_schedule`:

```
// set rotations to occur on Saturday at 00:00, but allow up to Sunday at 23:59
vault write database/static-roles/test-role \
    db_name=postgresql \
    username="vault" \
    rotation_schedule="0 0 * * SAT" \
    rotation_window="24h"
```

### CODE REVIEW

File: builtin/logical/database/backend.go [hunk 1]
```diff
@@ -127,6 +128,13 @@ func Backend(conf *logical.BackendConfig) *databaseBackend {
 	b.connections = syncmap.NewSyncMap[string, *dbPluginInstance]()
 	b.queueCtx, b.cancelQueueCtx = context.WithCancel(context.Background())
 	b.roleLocks = locksutil.CreateLocks()
+
+	// TODO(JM): don't allow seconds in production, this is helpful for
+	// development/testing though
+	parser := cron.NewParser(
+		cron.Second | cron.Minute | cron.Hour | cron.Dom | cron.Month | cron.DowOptional,
```
Comment (fairclothjm) [inline builtin/logical/database/backend.go:135]:
```suggestion
		cron.Second | cron.Minute | cron.Hour | cron.Dom | cron.Month | cron.Dow,
```

File: builtin/logical/database/path_roles.go [hunk 2]
```diff
@@ -623,14 +686,23 @@ func (b *databaseBackend) pathStaticRoleCreateUpdate(ctx context.Context, req *l
 		}
 	}
 
-	item.Priority = lvr.Add(role.StaticAccount.RotationPeriod).Unix()
+	_, rotationPeriodChanged := data.Raw["rotation_period"]
+	_, rotationScheduleChanged := data.Raw["rotation_schedule"]
```
Comment (austingebauer) [inline builtin/logical/database/path_roles.go:690]: Curious if there is a difference between these values from `data.Raw` and `rotationPeriodOk` / `rotationScheduleOk` from above?

File: builtin/logical/database/path_roles.go [hunk 3]
```diff
@@ -536,12 +559,18 @@ func (b *databaseBackend) pathStaticRoleCreateUpdate(ctx context.Context, req *l
 	}
 	role.StaticAccount.Username = username
 
-	// If it's a Create operation, both username and rotation_period must be included
-	rotationPeriodSecondsRaw, ok := data.GetOk("rotation_period")
-	if !ok && createRole {
-		return logical.ErrorResponse("rotation_period is required to create static accounts"), nil
+	rotationPeriodSecondsRaw, rotationPeriodOk := data.GetOk("rotation_period")
+	rotationSchedule := data.Get("rotation_schedule").(string)
+	rotationScheduleOk := rotationSchedule != ""
```
Comment (austingebauer) [inline builtin/logical/database/path_roles.go:564]: nit: I might be missing something but curious if there is a reason `GetOk` didn't work for `rotation_schedule`?

**Review by austingebauer — APPROVED** [review]
Looks good to me! Nice job on this @fairclothjm and @MilenaHC!

**Review by vinay-gopalan — APPROVED** [review]
LGTM! Had a couple questions :)

File: builtin/logical/database/path_roles.go [hunk 4]
```diff
@@ -536,12 +559,18 @@ func (b *databaseBackend) pathStaticRoleCreateUpdate(ctx context.Context, req *l
 	}
 	role.StaticAccount.Username = username
 
-	// If it's a Create operation, both username and rotation_period must be included
-	rotationPeriodSecondsRaw, ok := data.GetOk("rotation_period")
-	if !ok && createRole {
-		return logical.ErrorResponse("rotation_period is required to create static accounts"), nil
+	rotationPeriodSecondsRaw, rotationPeriodOk := data.GetOk("rotation_period")
+	rotationSchedule := data.Get("rotation_schedule").(string)
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply builtin/logical/database/path_roles.go:563]: Whoops! Just noticed after submitting the review that Austin had posted this question already. Disregard!

File: builtin/logical/database/backend.go [hunk 5]
```diff
@@ -127,6 +128,13 @@ func Backend(conf *logical.BackendConfig) *databaseBackend {
 	b.connections = syncmap.NewSyncMap[string, *dbPluginInstance]()
 	b.queueCtx, b.cancelQueueCtx = context.WithCancel(context.Background())
 	b.roleLocks = locksutil.CreateLocks()
+
+	// TODO(JM): don't allow seconds in production, this is helpful for
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply builtin/logical/database/backend.go:132]: Ah, also just saw Milena's [PR](https://github.com/hashicorp/vault/pull/22531) to only allow seconds in testing, so this comment should be resolved when that PR is merged 😅

File: builtin/logical/database/path_roles.go (hunk 3 above)
Comment (fairclothjm) [reply builtin/logical/database/path_roles.go:564]: I think this line was added a bit later when we realized we needed the boolean. Will update, thanks!

File: builtin/logical/database/path_roles.go (hunk 2 above)
Comment (fairclothjm) [reply builtin/logical/database/path_roles.go:690]: Nope, good catch!

**Comment by fairclothjm** [comment]
@austingebauer @vinay-gopalan Thanks for the reviews! I am going to address your comments in a follow up PR.

**Pull request merged**